moby/moby
# builder: warning if `yum` is attempted on Overlay

#### What did you do?

As already mentioned in #10180, `yum` does not work properly with the Overlay storage unless the `yum-plugin-ovl` package is installed.
This commit makes the builder print a warning if `yum` is attempted on the Overlay storage.

I also updated the doc and a PNG image, but I'm not sure how to edit the PNG image properly..
So please let me know whether there is the source SVG/PPT/ODP for the PNG image.

Off-topic: it seems that we need consideration for color-blinded persons.
<img src="https://cloud.githubusercontent.com/assets/9248427/13421680/313be98c-dfd3-11e5-8256-c52906c8b9e4.png" width="500" />
#### How did you do it?

`NewBuilder()` caches the result of `backend.SystemInfo()`.
`Builder.run()` prints a warning if `systemInfo.Driver` is `overlay` and `runConfig.Cmd` contains `yum`. The warning is printed only once.
#### How do I see it or verify it?

Prepare a Dockerfile like this:

``` dockerfile
FROM centos:centos6
RUN yum install -y sudo
RUN yum install -y perl
```

Make sure the `[Warning]` is printed (only when running with the Overlay driver).

```
$ docker build .
Sending build context to Docker daemon 2.048 kB
Step 1 : FROM centos:centos6
 ---> 2e4f3f04a056
Step 2 : RUN yum install -y sudo
 ---> Running in 9a001586afe7
 ---> [Warning] On Overlay, `yum` does not work properly unless the `yum-plugin-ovl` package is installed.For further information, please refer to https://docs.docker.com/engine/userguide/storagedriver/overlayfs-driver/
Loaded plugins: fastestmirror
Setting up Install Process
...
Installed:
  sudo.x86_64 0:1.8.6p3-20.el6_7                                                

Complete!
 ---> d721b5cbb456
Removing intermediate container 9a001586afe7
Step 3 : RUN yum install -y perl
 ---> Running in 1eade960f3ab
Loaded plugins: fastestmirror
Setting up Install Process
..
Running Transaction
  Installing : 1:perl-Pod-Escapes-1.04-141.el6_7.1.x86_64                   1/6 
  Installing : 1:perl-Module-Pluggable-3.90-141.el6_7.1.x86_64              2/6 
  Installing : 3:perl-version-0.77-141.el6_7.1.x86_64                       3/6 
  Installing : 4:perl-libs-5.10.1-141.el6_7.1.x86_64                        4/6 
  Installing : 1:perl-Pod-Simple-3.13-141.el6_7.1.x86_64                    5/6 
  Installing : 4:perl-5.10.1-141.el6_7.1.x86_64                             6/6

Rpmdb checksum is invalid: dCDPT(pkg checksums): perl-Pod-Simple.x86_64 1:3.13-141.el6_7.1 - u
```
#### A picture of a cute animal (not mandatory but encouraged)

https://upload.wikimedia.org/wikipedia/commons/f/f0/Emperor-pubs_hg.jpg
![emperor-pubs_hg](https://cloud.githubusercontent.com/assets/9248427/13421825/1c37b330-dfd4-11e5-89d3-711bd3b2e27e.jpg)

Update #10180

Signed-off-by: Akihiro Suda suda.akihiro@lab.ntt.co.jp
